lisp/ess-r-xref.el

```diff
@@ -0,0 +1,163 @@
+;;; ess-r-xref.el --- An xref backend for R. -*- lexical-binding: t -*-
+;;
+;; Author: Aaron Jacobs
+;; Created: 21 January 2018
+;; Maintainer: ESS-core <ESS-core@r-project.org>
+;;
+;; Keywords: languages, statistics, xref
+;; Package-Requires: ((emacs "25"))
+;;
+;; This file is part of ESS.
+;;
+;; This file is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation; either version 2, or (at your option)
+;; any later version.
+;;
+;; This file is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+;;
+;; A copy of the GNU General Public License is available at
+;; http://www.r-project.org/Licenses/
+
+;;; Commentary:
+
+;; This file contains an xref backend for `R-mode'.
+
+;;; Code:
+
+(require 'subr-x)
+(require 'xref)
+(require 'ess)
+
+
+;;; Xref API
+
+(defun ess-r-xref-backend ()
+  "An `xref-backend-functions' implementation for `R-mode'."
+  'ess-r)
+
+(cl-defmethod xref-backend-identifier-at-point ((_backend (eql ess-r)))
+  (ess-symbol-at-point))
+
+(cl-defmethod xref-backend-definitions ((_backend (eql ess-r)) symbol)
+  (inferior-ess-r-force)
+  (when (ess-r-xref--exists symbol)
+    (if (ess-r-xref--has-srcfile symbol)
+        (ess-r-xref--find-srcfile symbol)
+      (if (ess-r-xref--is-function symbol)
+          (ess-r-xref--find-body symbol)
+        ;; Ignore non-functions for now.
+        nil))))
+
+
+;;; Source File Locations
+
+(defun ess-r-xref--exists (symbol)
+  "Check whether SYMBOL is a known R symbol."
+  (ess-boolean-command (format "exists(\"%s\")\n" symbol)))
+
+(defun ess-r-xref--has-srcfile (symbol)
+  "Checks that the R symbol SYMBOL has a valid source file reference.
+
```

Comment:
A small stylistic comment. 

I am personally not a big fan of the recent trend in elisp comunity of splitting inline docs into single-sentence paragraphs. Especially if it's about internal functions. Code files are for code and as a developer you read the code 100 times but the docs only once or twice. Such docs take loads of space and their usability is questionable. Code should speak for itself IMO.

In this case you could safely keep only the last paragraph with no loss. Everyone can see that you are checking for srcref by glancing at the code.

Reply:
Ok.

Reply:
Removed the extra paragraph.